grails-app/services/org/pih/warehouse/inventory/ProductAvailabilityService.groovy

```diff
@@ -562,4 +566,81 @@ class ProductAvailabilityService {
         log.info "Updated ${results} product availability records for product ${product?.productCode}"
     }
 
+
+    /**
+     *
+     * @param commandInstance
+     * @return
+     */
+    List searchProducts(InventoryCommand command) {
+        def categories = inventoryService.getExplodedCategories([command.category])
+        List searchTerms = (command?.searchTerms ? Arrays.asList(command?.searchTerms?.split(" ")) : null)
+
+        // Only search if there are search terms otherwise the list of product IDs includes all products
+        def innerProductIds = !searchTerms ? [] : Product.createCriteria().list {
+            eq("active", true)
+            projections {
+                distinct 'id'
+            }
+            and {
+                searchTerms.each { searchTerm ->
+                    or {
+                        ilike("name", "%" + searchTerm + "%")
+                        inventoryItems {
+                            ilike("lotNumber", "%" + searchTerm + "%")
+                        }
+                    }
+                }
+            }
+        }
+
+        def products = Product.createCriteria().list {
+            eq("active", true)
+            and {
+                if (categories) {
+                    'in'("category", categories)
+                }
+                if (command.tags) {
+                    tags {
+                        'in'("id", command.tags*.id)
+                    }
+                }
+                if (command.catalogs) {
+                    productCatalogItems {
+                        productCatalog {
+                            'in'("id", command.catalogs*.id)
+                        }
+                    }
+                }
+                // This is pretty inefficient if the previous query does not narrow the results
+                if (innerProductIds) {
+                    'in'("id", innerProductIds)
+                }
+            }
+        }
+
+        def searchableTypes = ProductType.listAllBySupportedActivity([ProductActivityCode.SEARCHABLE])
+
+        def productsWithQoH = Product.executeQuery("""
```

Comment:
@jmiranda I left it done using executeQuery. I tried to quickly refactor it into criterias witth something like:

Product.createCriteria().list(max: command.maxResults, offset: command.offset) {
    projections {
        groupProperty("id", "id")
        productAvailabilities {
            sum("quantityOnHand", "quantityOnHand")
        }
    }
    eq("active", true)
    and {
        if (categories) {
            'in'("category", categories)
        }
        if (command.tags) {
            tags {
                'in'("id", command.tags*.id)
            }
        }
        if (command.catalogs) {
            productCatalogItems {
                productCatalog {
                    'in'("id", command.catalogs*.id)
                }
            }
        }
        // This is pretty inefficient if the previous query does not narrow the results
        if (innerProductIds) {
            'in'("id", innerProductIds)
        }

        productAvailabilities {
            'in'("location", command.location)
        }
        or {
            isNull("productType")
            and {
                'in'("productType", searchableTypes)
                productAvailabilities {
                    gt("quantityOnHand", 0)
                }
            }
        }
    }

The results were proper, but the grouping by product and summing the qoh was not working properly (or working at all). I tried different variants of projections, but nothing worked.

Reply:
(I can't get nice code structure here)

Reply:
You need three backticks.